### PR TITLE
Ocaml layer: enable ocaml-flycheck package via variable

### PR DESCRIPTION
**Pull request by actionshrimp**
This adds a layer variable to the ocaml layer to use flycheck.

The support was already in there, but it seems from discussion here https://github.com/syl20bnr/spacemacs/issues/9489 the recommended way of enabling it was editing the layer file manually to enable the package. This just uses the variable as a package toggle.

I also enabled flycheck automatically for ocaml's tuareg-mode as the optional package is loaded.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Do we really need this? I don't recall any other layer with flycheck support having such a variable? I'm pretty sure you can also just exclude flycheck-ocaml package to disable the support if you really want.

File: layers/+lang/ocaml/packages.el [hunk 1]
```diff
@@ -41,6 +40,8 @@
       :defer t
       :init
       (progn
+        (add-hook 'tuareg-mode-hook 'flycheck-mode)
```
Comment (rgrinberg) [inline layers/+lang/ocaml/packages.el]: I don't think this is necessary as we already have `(spacemacs/enable-flycheck 'tuareg-mode)`

Comment (actionshrimp) [reply layers/+lang/ocaml/packages.el]: Ah, it didnt seem to kick in but I'd missed that - will take another look and see if I can clean it up.

**Comment by actionshrimp** [comment]
Thanks for taking the time to take a look - the thing is that merlin mode (which is included) has it's own 'flycheck-esque' error reporting and error list which needs to be disabled for flycheck to work.

Perhaps it makes sense to use flycheck by default and allow switching back to the merlin error stuff by variable. But perhaps you're suggesting someone can just add flycheck-ocaml to their additional-packages list, in which case I'll change this to mention that in the README as it wasn't clear to me at first.

**Comment by rgrinberg** [comment]
You're right about merlin's own attempt to reinvent this. Luckily, we can disable it by setting `(setq merlin-error-after-save nil)`. Which is what we do after enabling flycheck-ocaml.

**Comment by actionshrimp** [comment]
Ah yep, sorry if I wasn't clear - I'm aware the layer does this stuff already I was just trying to make it more easy to turn in a way that didn't require manual edits of the layer source.

Turns out as flycheck wasnt in the layers package list (or rather it was, but was commented out) that `(spacemacs/enable-flycheck 'tuareg-mode)` wasn't kicking in, which I've now rectified and cleaned up.

I've also inverted the variable so that the current default behaviour (using merlin's error reporting) remains that way, but you can turn off merlin and 'revert' to the spacemacs standard of using flycheck for error reporting by setting the layer variable `ocaml-use-merlin-error-reporting` to `nil`.

**Comment by actionshrimp** [comment]
Aware this is getting a bit spammy now, so long story short: this is a workaround for these lines being commented out in `master` and `develop`:

https://github.com/syl20bnr/spacemacs/blob/develop/layers/+lang/ocaml/packages.el#L16-L17

being commented out which perhaps they shouldnt be but I thought perhaps they were for historical reasons. As you say if they were uncommented the user could just exclude `flycheck-ocaml` to disable it and revert to merlin.

**Comment by rgrinberg** [comment]
Hmm, I'm still not sure what problem is this variable solving.

* If the user doesn't list flycheck-ocaml in his packages, then merlin's own reporting should work.
* If the user lists flycheck-ocaml in his `dotspacemacs-additional-packages`, then spacemacs will correctly configure flycheck-ocaml and disable the merlin reporting.

We could argue about the default behavior here, but I don't see what option we are currently missing. Sorry if I'm being obtuse here.

**Comment by actionshrimp** [comment]
Ah - the problem is that adding `flycheck-ocaml` to `dotspacemacs-additional-packages` doesn't have any effect as the layer doesn't run the `ocaml/...init...` functions unless the packages are listed in the layer's package list.

**Comment by rgrinberg** [comment]
Dave Aitken <notifications@github.com> writes:

> Ah - the problem is that adding `flycheck-ocaml` to `dotspacemacs-additional-packages` doesn't have any effect as the layer doesn't run the `ocaml/...init...` functions unless the packages are listed in the layer's package list.

Hmm, well this mechanism works just fine for me. I have flycheck-ocaml
configured and working as expected and that's all I added to my config.

**Comment by actionshrimp** [comment]
I had a chance to test this in a bit more detail and I can't replicate what you observe on `develop` (also tried on `master`) with a clean `.emacs.d` and `.spacemacs` with the only changes as in the diff below:

```
$ diff --git a/Users/dave/.emacs.d/core/templates/.spacemacs.template b/Users/dave/.spacemacs
index 80e47f6f..f108b08a 100644
--- a/Users/dave/.emacs.d/core/templates/.spacemacs.template
+++ b/Users/dave/.spacemacs
@@ -51,8 +51,9 @@ This function should only modify configuration layer settings."
      ;;        shell-default-height 30
      ;;        shell-default-position 'bottom)
      ;; spell-checking
-     ;; syntax-checking
+     syntax-checking
      ;; version-control
+     ocaml
      )

    ;; List of additional packages that will be installed without being
@@ -62,7 +63,7 @@ This function should only modify configuration layer settings."
    ;; To use a local version of a package, use the `:location' property:
    ;; '(your-package :location "~/path/to/your-package/")
    ;; Also include the dependencies as they will not be resolved automatically.
-   dotspacemacs-additional-packages '()
+   dotspacemacs-additional-packages '(flycheck flycheck-ocaml)

    ;; A list of packages that cannot be updated.
    dotspacemacs-frozen-packages '()
```

I am fairly convinced from poking my own private layers that the layer mechanism uses that package list to decide which init functions to run - for example in the [https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Blang/elm/packages.el][elm layer] they include both flycheck and the elm flycheck package to achieve a similar thing.

**Comment by mattjbray** [comment]
I see the same as @actionshrimp (using latest `develop`). Adding `flycheck-ocaml` to `dotspacemacs-additional-packages` is not enough to trigger `ocaml/init-flycheck-ocaml`.

When I edit the `ocaml` layer to un-comment `flycheck-ocaml` in the `ocaml-packages` list, and restart Spacemacs in debug mode (`SPC q d`), then I see in the message log (`SPC w p m`):
```
(Spacemacs) flycheck-ocaml -> init (ocaml)...
```

So it does seem to me that some change in the layer is needed for users to enable `flycheck-ocaml` without having to edit the `ocaml` layer. I think a layer variable is perfect for this.

**Comment by rgrinberg** [comment]
I still don't see how such a variable is desirable. The elm layer that was pointed out doesn't include such a variable. Nor any other layer as far as I can tell.

If a user syntax-checking on and has flycheck installed, it's unlikely that he wants to tweak a bunch of layer specific variables as well? Is it possible that we just make flycheck-ocaml the default?

**Comment by actionshrimp** [comment]
Yeah that sounds good to me, just wasn't sure on the etiquette of swapping it over for ocaml merlin fans. I'll update the PR and put a note in the README that it can be disabled by adding to the excluded packages list which does seem to work when I get a sec, thanks.

**Comment by actionshrimp** [comment]
Should all be up to date now.

**Comment by syl20bnr** [comment]
Thanks to both of you for working on it ! 👍 
Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**